app/Jobs/ExtractZip.php

```diff
@@ -0,0 +1,181 @@
+		$this->history->owner_id = $this->userId;
+		$this->history->job = Str::limit('Extracting: ' . $this->originalBaseName, 200);
+		$this->history->parent_id = $this->albumID;
+		$this->history->status = JobStatus::READY;
+
+		$this->history->save();
+	}
+
+	/**
+	 * Execute the job.
+	 */
+	public function handle(): void
+	{
+		$extractedFolderName = $this->getExtractFolderName();
+
+		$pathExtracted = Storage::disk('extract-jobs')->path(date('Ymd') . $extractedFolderName);
+		$zip = new \ZipArchive();
+		if ($zip->open($this->filePath) === true) {
+			$zip->extractTo($pathExtracted);
+			$zip->close();
+
+			// clean up the zip file
+			unlink($this->filePath);
+
+			$this->history->status = JobStatus::SUCCESS;
+			$this->history->save();
+		} else {
+			throw new ZipExtractionException($this->filePath, $pathExtracted);
+		}
+
+		$newAlbum = $this->createAlbum($extractedFolderName, $this->albumID);
+		foreach (new \DirectoryIterator($pathExtracted) as $fileInfo) {
+			if ($fileInfo->isDot() || $fileInfo->isDir()) {
+				continue;
+			}
+
+			$extractedFile = new ExtractedJobFile($fileInfo->getRealPath(), $fileInfo->getFilename());
+			ProcessImageJob::dispatch($extractedFile, $newAlbum, $fileInfo->getMTime());
+		}
+	}
+
+	/**
+	 * Catch failures.
+	 *
+	 * @param \Throwable $th
+	 *
+	 * @return void
+	 */
+	public function failed(\Throwable $th): void
+	{
+		$this->history->status = JobStatus::FAILURE;
+		$this->history->save();
+
+		if ($th->getCode() === 999) {
+			$this->release();
+		} else {
+			Log::error(__LINE__ . ':' . __FILE__ . ' ' . $th->getMessage(), $th->getTrace());
+		}
+	}
+
+	/**
+	 * Given a name and parent we create it.
+	 *
+	 * @param string      $newAlbumName
+	 * @param string|null $parentID
+	 *
+	 * @return Album new album
+	 */
+	private function createAlbum(string $newAlbumName, ?string $parentID): Album
+	{
+		if (SmartAlbumType::tryFrom($parentID) !== null) {
+			$parentID = null;
+		}
+
+		/** @var Album $parentAlbum */
+		$parentAlbum = $parentID !== null ? Album::query()->findOrFail($parentID) : null; // in case no ID provided -> import to root folder
+		$createAlbum = new Create($this->userId);
+
+		return $createAlbum->create($this->prepareAlbumName($newAlbumName), $parentAlbum);
+	}
+
+	/**
+	 * Todo Later: add renamer module.
+	 *
+	 * @param string $albumNameCandidate
+	 *
+	 * @return string
+	 */
+	private function prepareAlbumName(string $albumNameCandidate): string
+	{
+		return trim(str_replace('_', ' ', $albumNameCandidate));
+	}
+
+	/**
+	 * Returns a folder name where:
+	 * - spaces are replaced by `_`
+	 * - if folder already exists (with date prefix) then we pad with _(xx) where xx is the next available number.
+	 *
+	 * @return string
+	 */
+	private function getExtractFolderName(): string
+	{
+		$baseNameWithoutExtension = substr($this->originalBaseName, 0, -4);
+
+		// Save that one (is default if no existing folder found).
+		$orignalName = str_replace(' ', '_', $baseNameWithoutExtension);
```

Comment:
Why are you doing this replacement here? Does ZipArchive not handle spaces correctly or something?

Reply:
It may, but I don't trust spaces in paths.
And given that we replace '_' by ' ' later in file names. I figured it would be idem potent.